File: files/CH-103.R
library(tidyverse)
library(readxl)

path = "files/CH-103 Custom Average.xlsx"
input = read_excel(path, range = "B2:D21")
test = read_excel(path, range = "I2:J6")

today = as.POSIXct(as.Date("2024-08-20"))

result = input %>%
  mutate(del_time = as.numeric(`Delivery Date` - `Order Date`),
         adj_del_time = ifelse(is.na(`Delivery Date`), today - `Order Date`, del_time)) %>%
  mutate(avg_del_time = mean(del_time, na.rm = T), .by = `Product ID`) %>%
  filter(!is.na(del_time) | adj_del_time >= avg_del_time) %>%
  summarise(`Avg delivery time` = mean(adj_del_time, na.rm = T), .by = `Product ID`) %>%
  arrange(`Product ID`)

identical(result$`Avg delivery time`, test$`Avg delivery time`)
#> [1] TRUE

File: files/CH-103.py
import pandas as pd
import numpy as np

path = "CH-103 Custom Average.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows = 1, nrows = 19)
test = pd.read_excel(path, usecols="I:J", skiprows = 1, nrows = 4)

today = pd.to_datetime("2024-08-20")

input["Delivery Time"] = (input["Delivery Date"] - input["Order Date"]).dt.days
input["Adjusted Delivery Time"] = np.where(input["Delivery Date"].isnull(), (today - input["Order Date"]).dt.days, input["Delivery Time"])
input = input[input["Delivery Time"].notnull() | (input["Adjusted Delivery Time"] > input.groupby("Product ID")["Delivery Time"].transform("mean"))]\
    .groupby("Product ID")["Adjusted Delivery Time"].mean().reset_index()

print(input["Adjusted Delivery Time"].equals(test["Avg delivery time"])) # True